src/preamble.js

```diff
@@ -897,6 +897,25 @@ if (ENVIRONMENT_IS_PTHREAD) {
 }
 #endif
 
+#if STACK_OVERFLOW_CHECK
+// Initializes the stack cookie. Called at the startup of main and at the startup of each thread in pthreads mode.
+function writeStackCookie() {
+  assert((STACK_MAX & 3) == 0);
+  HEAPU32[(STACK_MAX >> 2)-1] = 0x02135467;
+  HEAPU32[(STACK_MAX >> 2)-2] = 0x89BACDFE;
+}
+
+function checkStackCookie() {
+  if (HEAPU32[(STACK_MAX >> 2)-1] != 0x02135467 || HEAPU32[(STACK_MAX >> 2)-2] != 0x89BACDFE) {
+    abort('Stack overflow! Stack cookie has been overwritten, expected hex dwords 0x89BACDFE and 0x02135467, but received 0x' + HEAPU32[(STACK_MAX >> 2)-2].toString(16) + ' ' + HEAPU32[(STACK_MAX >> 2)-1].toString(16));
+  }
+}
+
+function abortStackOverflow(allocSize) {
+  abort('Stack overflow! Attempted to allocate ' + allocSize + ' bytes on the stack, but stack has only ' + (STACK_MAX - STACKTOP) + ' bytes available!');
```

Comment:
Thanks! I was thinking about this, and wrote the test to confirm, and shrugged when it looked ok there. But of course my test was wrong, so updated that.

I remember getting this silently wrong way back when initially implementing stacks for pthreads as well, and I think we should fix this and not leave stale variables around. Posted #4146 as one way to do it, how does that look?